docs/_includes/Types.md

```diff
@@ -456,49 +456,107 @@ GenericParameters =
+must be restricted in a particular way, in which case Dafny
+will add the appropriate suffix, such as `(==)`, automatically.
+
+### Equality-supporting type parameters: `T(==)`
+
+Designating a type parameter with the `(==)` suffix indicates that
+the parameter may only be replaced in non-ghost contexts
+with types that are known to
+support run-time equality comparisons (`==` and `!=`).
+All types support equality in ghost contexts,
+as if, for some types, the equality function is ghost.
+
+For example,
 ```dafny
 method Compare<T(==)>(a: T, b: T) returns (eq: bool)
 {
   if a == b { eq := true; } else { eq := false; }
 }
 ```
 is a method whose type parameter is restricted to equality-supporting
-types.  Again, note that _all_ types support equality in _ghost_
+types when used in a non-ghost context.
+Again, note that _all_ types support equality in _ghost_
 contexts; the difference is only for non-ghost (that is, compiled)
-code.  Co-inductive datatypes, function types, as well as inductive
+code.  Co-inductive datatypes, function types, and inductive
 datatypes with ghost parameters are examples of types that are not
 equality supporting.
 
-TO BE WRITTEN: There are now more such modes
+### Auto-initializable types: `T(0)`
 
-[^fn-type-mode]:  Being equality-supporting is just one of many
-    _modes_ that one can imagine types in a rich type system to have.
-    For example, other modes could include having a total order,
-    being zero-initializable, and possibly being uninhabited.  If
-    Dafny were to support more modes in the future, the `( )`-suffix
-    syntax may be extended.  For now, the suffix can only indicate the
-    equality-supporting mode.
+All Dafny variables of a given type hold a legal value of that type;
+if no explicit initialization is given, then an arbitrary value is
+assumed by the verifier and supplied by the compiler,
+ that is, the variable is _auto-initialized_.
+During verification, this means that any subsequent uses of that
+variable must hold for any value.
+For example,
+```dafny
+method m() {
+  var n: nat; // Initialized to arbitrary value
+  assert n >= 0; // true, regardless of the value of n
+  var i: int;
+  assert i >= 0; // possibly false, arbitrary ints may be negative
+}
+```
 
-Dafny has some inference support that makes certain signatures less
-cluttered (described in [Section Type Inference](#sec-type-inference)).
-In some cases, this support will
-infer that a type parameter must be restricted to equality-supporting
-types, in which case Dafny adds the "`(==)`" automatically.
+However, by default, non-ghost computations must follow definite-assignment
+rules, in which case local variables must be assigned a value before use.
+For some types, the compiler can choose an initial value, but for others
+it does not; for more details see the section describing [Definite Assignment](#sec-definite-assignment) and the `-definiteAssignment` command-line option.
```

Comment:
The first sentence here is not right (perhaps because the `/help` message is not clear enough). I suggest replacing both sentences with:

For some types, the compiler can choose an initial value, but for others it does not.
Variables and fields of a type that the compiler does not auto-initialize are subject to
_definite-assignment rules_. These ensure that the program explicitly assigns the
variable before it is used. For more details, see the section describing
[Definite Assignment](#sec-definite-assignment) and the `-definiteAssignment` command-line
option.

Also, the `-help` description of `-definiteAssignment:1` can be clarified to:

    1 (default) - enforces definite-assignment rules for variables and fields
                  of types that do not support auto-initialization

Reply:
Done